### ISSUE TITLE
Support for Hibernate 6

### ISSUE THREAD

**Issue by odrotbohm** [querydsl/querydsl#3166]
Hibernate has released a CR for their 6th generation. While trying to [upgrade (or at least support it in) Spring Data JPA](https://github.com/spring-projects/spring-data-jpa/issues/2423), our tests fail with a `NullPointerException` coming from `JPAUtil.setConstants(…)`:

```
[ERROR] Failures: 
[ERROR]   QuerydslJpaPredicateExecutorUnitTests.findByFluentPredicateOneValue:369 
Expecting actual throwable to be an instance of:
  org.springframework.dao.IncorrectResultSizeDataAccessException
but was:
  java.lang.NullPointerException: Cannot invoke "org.hibernate.query.BindableType.getBindableJavaType()" because "this.anticipatedType" is null
	at org.hibernate.query.spi.AbstractQueryParameter.getParameterType(AbstractQueryParameter.java:61)
	at com.querydsl.jpa.impl.JPAUtil.setConstants(JPAUtil.java:53)
	at com.querydsl.jpa.impl.AbstractJPAQuery.createQuery(AbstractJPAQuery.java:133)
	...(79 remaining lines not displayed - this can be changed with Assertions.setMaxStackTraceElementsDisplayed)
[ERROR]   QuerydslJpaPredicateExecutorUnitTests.findOneWithPredicateThrowsExceptionForNonUniqueResults:329 
Expecting actual throwable to be an instance of:
  org.springframework.dao.IncorrectResultSizeDataAccessException
but was:
  java.lang.NullPointerException: Cannot invoke "org.hibernate.query.BindableType.getBindableJavaType()" because "this.anticipatedType" is null
	at org.hibernate.query.spi.AbstractQueryParameter.getParameterType(AbstractQueryParameter.java:61)
	at com.querydsl.jpa.impl.JPAUtil.setConstants(JPAUtil.java:53)
	at com.querydsl.jpa.impl.AbstractJPAQuery.createQuery(AbstractJPAQuery.java:133)
	...(76 remaining lines not displayed - this can be changed with Assertions.setMaxStackTraceElementsDisplayed)
```

**Comment by odrotbohm** [comment]
Actually, it looks like the `NPE` is caused by Hibernate in the first place.

**Comment by odrotbohm** [comment]
There's a reproducer to be found [here](https://github.com/odrotbohm/hibernate-bugs/tree/main/h6-querydsl) (plain Querydsl + Hibernate 6). It looks like the issue only appears when certain types of predicates are involved. `contains` triggers the issue, `eq` does not. I still think it's a bug in Hibernate 6 as it works just fine on 5.6.5. That said, I think *you* need to bring this up with Hibernate as I cannot argue on behalf of Querydsl.

**Comment by odrotbohm** [comment]
It looks like the issue has been resolved in Hibernate 6.0 CR2.

**Comment by MSDehghan** [comment]
Just to note, The final version of Hibernate 6.0 is released.
https://in.relation.to/2022/03/31/orm-60-final/
